pkg/auth/auth.go

```diff
@@ -203,23 +211,29 @@ func parseCredentialsKey(arg string, acceptRepositories bool) (key, registry str
 }
 
 // getRegistryName scrubs and parses the input to get the server name
-func getRegistryName(server string) string {
-	// removes 'http://' or 'https://' from the front of the
-	// server/registry string if either is there.  This will be mostly used
-	// for user input from 'Buildah login' and 'Buildah logout'.
-	server = trimScheme(server)
+func getRegistryName(server string) (string, error) {
+	server, err := trimScheme(server)
+	if err != nil {
+		return "", err
+	}
 	// gets the registry from the input. If the input is of the form
 	// quay.io/myuser/myimage, it will parse it and just return quay.io
 	split := strings.Split(server, "/")
-	return split[0]
+	return split[0], nil
 }
 
-// trimScheme removes the HTTP(s) scheme from the provided repository.
-func trimScheme(repository string) string {
-	// removes 'http://' or 'https://' from the front of the
-	// server/registry string if either is there.  This will be mostly used
-	// for user input from 'Buildah login' and 'Buildah logout'.
-	return strings.TrimPrefix(strings.TrimPrefix(repository, "https://"), "http://")
+// trimScheme removes the http{s} scheme from the provided repository and
+// returns the host[:port] parts only.  The repository is returned as is if
+// there is no http{s} prefix.
+func trimScheme(repository string) (string, error) {
+	if !strings.HasPrefix(repository, "https://") && !strings.HasPrefix(repository, "http://") {
+		return repository, nil
+	}
+	u, err := url.Parse(repository)
+	if err != nil {
+		return "", fmt.Errorf("trimming http{s} prefix: %v", err)
+	}
+	return u.Host, nil
```

Comment:
So this function is now `replaceURLByHostPort`, and not just `trimScheme` (but non-scheme-prefixed values are returned unchanged, including the namespaces)?

That… works but feels quite hard to follow. At the very least, please rename this function appropriately.

---

I think I’d prefer just explicitly saying that “URL-like values are a special case” and inlining `trimScheme` at the top of `parseCredentialsKey` (if `… HasPrefix(arg) … { url.Parse … arg = url.Host }`); then `getRegistryName` would not need to call `trimScheme`.

(Even `getRegistryName` could then be moved so that it is only called once, always, and inlined as well.)

Reply:
Great ideas!